CLARIN-PL/doccano
# Feat: BE - Add new endpoint to create user, quick fix for the annotation starting time logging

Hello, this PR covers 2 things:
- Create new api for create user (currently, we don't have the endpoint for it, we have to create user by using django admin page. But to automatically create multiple users, I need it. Then I added it)
- Quick fix for the `ExampleAnnotateStartState` (there was an issue that for  eg: the annotator click to the start annotation button to start the work on example 1, then the start time of example 1 will be logged. But if the annotator doesn't finish annotating that example, he logout or leaves the annotation page, and then come back later by clicking on the start annotation button again. Then currently, the mechanism of ExampleAnnotateStartState table doesn't accept the duplication and will delete it. To fix that, I change it to be updated. So every time an example is starting again by the same annotator, it will be updated the starting time).

-------
What's changed:
- `backend/config/settings/base.py`, `backend/poetry.lock`: added new library for create user endpoint
- `backend/users/views.py`: added new view for create user
- `backend/users/urls.py`: added new endpoint for creating user
- `backend/users/tests/test_views.py`, `backend/users/tests/utils.py`: added test case for endpoint create_user (and the utils function for testing)
- `backend/examples/views/example_state.py`: fix for the `ExampleAnnotateStartState`, replace `delete()` by `update()`
- `backend/labels/tests/test_text_label.py`: update test cases for `TextLabel` due to some changes the `TextLabel` constraints previously

Screenshots:
The `ExampleAnnotateStartState` (starting annotation time for multiple annotators)
![Screenshot from 2022-10-28 17-13-46](https://user-images.githubusercontent.com/17596118/198672411-d7d9c740-7b82-4c92-bad5-d757e8e1df8a.png)

New created users using create_user api (I called the api from client to create them automatically)
![Screenshot from 2022-10-28 17-15-32](https://user-images.githubusercontent.com/17596118/198672887-0cd2d564-b404-4fa0-bb85-607a4269cd3f.png)

Edit: I just remember that the label_config files in the backend are not needed now, since in frontend we have the auto import and from doccano-client we will have that import soon (hopefully today :sweat_smile:) so I removed them :smile:  

Edit 2: Hello, just note here that this pr is related to this branch https://github.com/haanh764/doccano-client/tree/new_project_type - I have adjusted the doccano-client to be able to work with our changes in the doccano backend.
What's done are:
- Create users (given a file contains usernames and passwords)
- Create projects (for each dataset we will have 5 projects corresponded to 5 annotation modes)
- Upload dataset to each project
- Upload corresponded labels_config for each project
- Assign member for each project
How to test it:
- Clone the repo doccano-client, checkout to the branch above
- Run `poetry install` to install the requirements
- Change directory to `/demo`
- Run script for seeing the demo `python project_creator.py`

You can see a quick demo here
https://user-images.githubusercontent.com/17596118/198699954-15a47108-c2d6-466a-9ea7-b40513ecbe4c.mp4
